File: BinaryTreeZigzagLevelOrderTraversal/BinaryTreeZigzagLevelOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
    	vector<vector<int>> result;
    	bool flag = true;
    	buildZZTree(root, result, flag, 0);
    	return result;    
    }
private:
	void buildZZTree(TreeNode* node, vector<vector<int>> &result, bool flag, int depth) {
		if(node == nullptr)
			return;
		if(result.size() <= depth)
			result.push_back(vector<int>(0,0));
		if(flag == false)
			result[depth].insert(result[depth].begin(), node->val);
		else
			result[depth].push_back(node->val);
		buildZZTree(node->left, result, !flag, depth + 1);
		buildZZTree(node->right, result, !flag, depth + 1);
		return;
	}

};

File: BinaryTreeZigzagLevelOrderTraversal/BinaryTreeZigzagLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        res = []
        flag = True
        buildZZTree(root, res, flag, 0)
        return res

def buildZZTree(root, res, flag, depth):
    if root is None:
        return
    if len(res) <= depth:
        res.append([])
    if not flag:
        res[depth].insert(0, root.val)
    else:
        res[depth].append(root.val)
    buildZZTree(root.left, res, not flag, depth + 1)
    buildZZTree(root.right, res, not flag, depth + 1)
    return res